services/sms/js/app.js

```diff
@@ -0,0 +1,163 @@
+(function(window) {
+  'use strict';
+
+  function debug(str) {
+    console.log('SMSService -*-:' + str);
+  }
+
+  debug('/services/sms/js/app.js --> loading!!');
+
+  // Ok, this kinda sucks because most APIs (and sms is one of them) cannot
+  // be accessed from outside the main thread. So basically everything has to go
+  // down to the SW thread, then back up here for processing, then back down to
+  // be sent to the client. Yay us!
+  var _locks = {};
+  var _observers = {};
+
+  var _sms = navigator.mozMobileMessage;
+
+  function buildDOMCursorAnswer(operation, channel, request) {
+    var remotePortId = request.remotePortId;
+    // Params for the local operation:
+    var opData = request.remoteData.data.params || [];
+    var reqId = request.remoteData.id;
+
+    // FIX-ME: Due to the way FakeDOMCursorRequest is implemented, we
+    // have to return all the fetched data on a single message
+    var cursor = _sms[operation](...opData);
+    var _messages = [];
+    cursor.onsuccess = function onsuccess() {
+      debug(operation + '.cursor.onsuccess: ' + this.done + ', ' +
+            JSON.stringify(this.result));
+      if (!this.done) {
+        _messages.push(window.ServiceHelper.cloneObject(this.result));
+        this.continue();
```

Comment:
I don't get this... if this.done is false then this.result should not be undefined... this.result should be undefined only when this.done is true, and then we don't push the element. Isn't it so?